### PR TITLE
[MODCON-38] - Process USER_UPDATED event to update username if changed

### PR DESCRIPTION
**Pull request by Dmitriy-Butramyou**
[MODCON-38](https://issues.folio.org/browse/MODCON-38)

## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java [hunk 1]
```diff
@@ -78,6 +79,44 @@ public void createPrimaryUserAffiliation(String eventPayload) {
...
+
+      var consortiaTenant = tenantService.getByTenantId(userEvent.getTenantId());
+      if (consortiaTenant == null) {
+        log.warn("Tenant {} not exists in consortia", userEvent.getTenantId());
+        return;
+      }
+      UUID userId = getUserId(userEvent);
+      String newUsername = userEvent.getUserDto().getUsername();
+
+      UserTenantCollection userTenantCollection = userTenantService.getByUserId(consortiaTenant.getConsortiumId(), userId, 0, Integer.MAX_VALUE);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:100]: We should use method: findByUserIdAndTenantId, it already exists in DAO, need to add it it UseTenantService and use ,
tenantId we should use from eventPayload

File: src/main/java/org/folio/consortia/repository/UserTenantRepository.java [hunk 2]
```diff
@@ -35,6 +35,10 @@ public interface UserTenantRepository extends JpaRepository<UserTenantEntity, UU
   @Query("DELETE FROM UserTenantEntity ut WHERE ut.userId= ?1 AND ut.isPrimary= true")
   void deleteByUserIdAndIsPrimaryTrue(UUID userId);
 
+  @Modifying
+  @Query("UPDATE UserTenantEntity ut SET ut.username= ?1 WHERE ut.userId= ?2")
+  void setUsernameByUserId(String username, UUID userId);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/repository/UserTenantRepository.java:40]: We don't have logic to update username of shadow users now, we are updating only username of real user, the same we should do in mod-consortia, so we need to add tenantId param and user it in WHERE filtering as well.

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java [hunk 3]
```diff
@@ -78,6 +79,44 @@ public void createPrimaryUserAffiliation(String eventPayload) {
...
+      boolean isUsernameChanged = userTenantCollection.getUserTenants().stream()
+        .anyMatch(userTenant -> ObjectUtils.notEqual(userTenant.getUsername(), newUsername));
+
+      if (isUsernameChanged) {
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:104]: Just help for reviewers, we are not storing email, phoneNumber, mobilePhoneNumber in mod-consortia, so username is only field that can be updated.
But we should send all fields for CONSORTIUM_PRIMARY_AFFILIATION events to populate in mod-users central tenant

File: src/main/resources/db/changelog/changes/create-user-tenant-table.xml [hunk 4]
```diff
@@ -55,4 +55,10 @@
       <column name="updated_date" type="timestamp without time zone"/>
     </addColumn>
   </changeSet>
+
+  <changeSet id="MODCON-38@@add-central-tenant-id-field" author="d.butramyou">
```
Comment (SerhiiNosko) [inline src/main/resources/db/changelog/changes/create-user-tenant-table.xml:59]: We don't need this change, we are storing centralTenantId in table consortia-configuration and in tenant table we have flag isCentral. So we just need to have centralTenantId in payload to send to mod-users central tenant, but not in mod-consortia DB

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java [hunk 5]
```diff
@@ -78,6 +79,44 @@ public void createPrimaryUserAffiliation(String eventPayload) {
     }
   }
 
+  @Override
+  @SneakyThrows
+  @Transactional
+  public void updatePrimaryUserAffiliation(String eventPayload) {
+    FolioExecutionContext currentContext = (FolioExecutionContext) folioExecutionContext.getInstance();
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:86]: no need to have this variable, setting context is working with a plain folioExecutionContext field

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java [hunk 6]
```diff
@@ -78,6 +79,44 @@ public void createPrimaryUserAffiliation(String eventPayload) {
...
+
+      if (isUsernameChanged) {
+        userTenantService.updateUsernameInPrimaryUserTenantAffiliation(userId, newUsername);
+        log.info("Username in primary affiliation has been updated for the user: {}", userEvent.getUserDto().getId());
+      }
+
+      PrimaryAffiliationEvent affiliationEvent = createPrimaryAffiliationEvent(userEvent);
+      String data = objectMapper.writeValueAsString(affiliationEvent);
+
+      try (var context = new FolioExecutionContextSetter(prepareContextForTenant(centralTenantId, folioModuleMetadata, currentContext))) {
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:112]: isn't it central tenant folioExecutionContext already?

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java (hunk 1 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java]: Refactored

File: src/main/java/org/folio/consortia/repository/UserTenantRepository.java (hunk 2 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/repository/UserTenantRepository.java]: Added

File: src/main/resources/db/changelog/changes/create-user-tenant-table.xml (hunk 4 above)
Comment (Dmitriy-Butramyou) [reply src/main/resources/db/changelog/changes/create-user-tenant-table.xml]: Deleted

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java (hunk 5 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java]: Deleted

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java (hunk 6 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java]: Refactored

File: src/main/resources/swagger.api/schemas/userTenant.yaml [hunk 7]
```diff
@@ -16,6 +16,8 @@ UserTenant:
       type: string
     isPrimary:
       type: boolean
+    centralTenantId:
+      type: string
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply src/main/resources/swagger.api/schemas/userTenant.yaml:20]: Added

File: src/main/java/org/folio/consortia/domain/entity/UserTenantEntity.java [hunk 8]
```diff
@@ -31,6 +31,7 @@ public class UserTenantEntity extends AuditableEntity {
   private TenantEntity tenant;
 
   private Boolean isPrimary;
+  private String centralTenantId;
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/domain/entity/UserTenantEntity.java:34]: This field is not needed in user-tenant entity

Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/domain/entity/UserTenantEntity.java]: deleted

**Review by SerhiiNosko — APPROVED** [review]

**Review by tatsiana-tarhonskaya — APPROVED** [review]

**Comment by SerhiiNosko** [comment]
@Dmitriy-Butramyou I forgot that we need to support existed users, we need populate email, phoneNumber and mobilePhoneNumber for them as well, so we need adjust this payload to include these 2 fields:
https://github.com/folio-org/mod-consortia/blob/master/src/main/java/org/folio/consortia/service/impl/SyncPrimaryAffiliationServiceImpl.java#L103 
I also will update story with this findings

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java [hunk 9]
```diff
@@ -78,6 +81,39 @@ public void createPrimaryUserAffiliation(String eventPayload) {
...
+
+      UserTenantEntity userTenant = userTenantService.getByUserIdAndTenantId(userId, userEvent.getTenantId());
+      boolean isUsernameChanged = ObjectUtils.notEqual(userTenant.getUsername(), newUsername);
+
+      if (isUsernameChanged) {
+        userTenantService.updateUsernameInPrimaryUserTenantAffiliation(userId, newUsername, userEvent.getTenantId());
+        log.info("Username in primary affiliation has been updated for the user: {}", userEvent.getUserDto().getId());
+      }
+
+      PrimaryAffiliationEvent affiliationEvent = createPrimaryAffiliationEvent(userEvent);
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:108]: PrimaryAffiliationEvent should also include email, phoneNumber, mobilePhoneNumber, centralTenantId

**Review by azizbekxm — APPROVED** [review]

File: src/main/java/org/folio/consortia/service/impl/SyncPrimaryAffiliationServiceImpl.java [hunk 10]
```diff
@@ -119,6 +119,9 @@ private PrimaryAffiliationEvent createPrimaryAffiliationEvent(SyncUser user, Str
     event.setUserId(UUID.fromString(user.getId()));
     event.setUsername(user.getUsername());
     event.setTenantId(tenantId);
+    event.setEmail(user.getEmail());
+    event.setPhoneNumber(user.getPhoneNumber());
+    event.setMobilePhoneNumber(user.getMobilePhoneNumber());
```
[the remark this answers is not in the document]
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/service/impl/SyncPrimaryAffiliationServiceImpl.java:124]: added

File: src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java (hunk 9 above)
Comment (Dmitriy-Butramyou) [reply src/main/java/org/folio/consortia/service/impl/UserAffiliationServiceImpl.java:108]: added

**Pull request merged**